### ISSUE TITLE
 pm list packages io.appium.uiautomator2.server.test' timed out after 20000ms'; Stderr: ''; Code: 'null'

### ISSUE THREAD

**Issue by chunyuong** [appium/appium#9834]
I have installed  io.appium.uiautomator2.server.test and io.appium.uiautomator2.server.test;The timeout problems were found ,this is log:

[debug] [ADB] Running 'D:\Program Files\android-sdk-windows\platform-tools\adb.exe' with args: ["-P",5037,"-s","84a0aa5d","shell","pm","list","packages","io.appium.uiautomator2.server.test"]
[debug] [ADB] Running 'D:\Program Files\android-sdk-windows\platform-tools\adb.exe' with args: ["-P",5037,"-s","84a0aa5d","shell","pm","list","packages","io.appium.uiautomator2.server.test"]
[ADB] Error: Error finding if app is installed. Original error: Error executing adbExec. Original error: 'Command ''D:\\Program Files\\android-sdk-windows\\platform-tools\\adb.exe' -P 5037 -s 84a0aa5d shell pm list packages io.appium.uiautomator2.server.test' timed out after 20000ms'; Stderr: ''; Code: 'null'
    at Object.wrappedLogger.errorAndThrow (../../lib/logging.js:63:13)
    at ADB.callee$0$0$ (../../../lib/tools/apk-utils.js:28:9)
    at tryCatch (C:\Users\liuchunyong\AppData\Local\Programs\appium-desktop\resources\app\node_modules\appium\node_modules\babel-runtime\regenerator\runtime.js:67:40)
    at GeneratorFunctionPrototype.invoke [as _invoke] (C:\Users\liuchunyong\AppData\Local\Programs\appium-desktop\resources\app\node_modules\appium\node_modules\babel-runtime\regenerator\runtime.js:315:22)
    at GeneratorFunctionPrototype.prototype.(anonymous function) [as throw] (C:\Users\liuchunyong\AppData\Local\Programs\appium-desktop\resources\app\node_modules\appium\node_modules\babel-runtime\regenerator\runtime.js:100:21)
    at GeneratorFunctionPrototype.invoke (C:\Users\liuchunyong\AppData\Local\Programs\appium-desktop\resources\app\node_modules\appium\node_modules\babel-runtime\regenerator\runtime.js:136:37)
 Error: Error finding if app is installed. Original error: Error executing adbExec. Original error: 'Command ''D:\\Program Files\\android-sdk-windows\\platform-tools\\adb.exe' -P 5037 -s 84a0aa5d shell pm list packages io.appium.uiautomator2.server.test' timed out after 20000ms'; Stderr: ''; Code: 'null'
    at Object.wrappedLogger.errorAndThrow (../../lib/logging.js:63:13)
    at ADB.callee$0$0$ (../../../lib/tools/apk-utils.js:28:9)
    at tryCatch (C:\Users\liuchunyong\AppData\Local\Programs\appium-desktop\resources\app\node_modules\appium\node_modules\babel-runtime\regenerator\runtime.js:67:40)
    at GeneratorFunctionPrototype.invoke [as _invoke] (C:\Users\liuchunyong\AppData\Local\Programs\appium-desktop\resources\app\node_modules\appium\node_modules\babel-runtime\regenerator\runtime.js:315:22)
    at GeneratorFunctionPrototype.prototype.(anonymous function) [as throw] (C:\Users\liuchunyong\AppData\Local\Programs\appium-desktop\resources\app\node_modules\appium\node_modules\babel-runtime\regenerator\runtime.js:100:21)
    at GeneratorFunctionPrototype.invoke (C:\Users\liuchunyong\AppData\Local\Programs\appium-desktop\resources\app\node_modules\appium\node_modules\babel-runtime\regenerator\runtime.js:136:37)

**Comment by chunyuong** [comment]
Test phone: HUAWEI  VTR-TL00

**Comment by mykola-mokhnach** [comment]
This looks like a problem with adb/device under test itself. Try another device.

**Comment by chunyuong** [comment]
@mykola-mokhnach   I  try it “adb shell pm list packages io.qppium”  , I found that the command was stuck all the time ,As shown in the following figure:
![image](https://user-images.githubusercontent.com/26139195/34294153-2666a560-e742-11e7-8339-00049f2bc29e.png)

**Comment by mykola-mokhnach** [comment]
Then the problem is for sure not on appium side.

**Comment by speesystack-chentianyu** [comment]
@chunyuong
 I ran into the same problem days ago. In my case, the command did get stuck, but not in a way like forever stuck. Instead, it would first print out the packages, then hang there doing nothing for around two minutes before finally gets terminated. Since Appium sets a timeout of 20000 ms for those adb commands, running the commands will mostly result with a time-out exception.

I have done some experiments and found a few facts.

1. The commands will only stuck if run from adb. If you run `adb shell` first, then type `pm list packages` from within the shell, the second command returns immediately instead of getting stuck.
2. Only commands related to package management will get stuck.

As my guess, some security related apps (360Safe for example) prevents `pm` or `shell` from sending termination signals to `adb`. 


So, there are several solutions to this problem.

1. Modify the `appium-adb` module. Change the default timeout for `adb` commands to 10 minutes, then recompile that module. This works but each time you start a new session it takes a really long time. 
2. Remove any suspicious app in your device. In my case, I got lucky and successfully flashed in a clean ROM. After that, those adb commands no longer get stuck. I believe the problem is caused by a custom version of 360Safe app which is installed in the device (ChinaMobile, A3s Model 653) by the manufacturer.
3. Buy another device with a safe and clean ROM.
